src/stream/src/executor/backfill/cdc/cdc_backfill.rs

```diff
@@ -168,29 +175,107 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
         // if not, we should bypass the backfill directly.
         let mut state_impl = self.state_impl;
 
-        let mut upstream = transform_upstream(upstream, &self.output_columns)
-            .boxed()
-            .peekable();
-
         state_impl.init_epoch(first_barrier_epoch).await?;
 
         // restore backfill state
         let state = state_impl.restore_state().await?;
         current_pk_pos = state.current_pk_pos.clone();
 
-        let to_backfill = !self.options.disable_backfill && !state.is_finished;
+        let need_backfill = !self.options.disable_backfill && !state.is_finished;
 
         // Keep track of rows from the snapshot.
         let mut total_snapshot_row_count = state.row_count as u64;
 
+        // After init the state table and forward the initial barrier to downstream,
+        // we now try to create the table reader with retry.
+        // If backfill hasn't finished, we can ignore upstream cdc events before we create the table reader;
+        // If backfill is finished, we should forward the upstream cdc events to downstream.
+        let mut upstream_buffer = UpstreamMessageBuffer::new(upstream);
+        let table_reader: Option<ExternalTableReaderImpl>;
+        let mut backoff = get_backoff_strategy();
+        loop {
+            let create_result: StreamExecutorResult<_> = try {
+                let reader = upstream_buffer
+                    .run_future(create_table_reader(&self.external_table))
+                    .await?;
+                reader
+            };
+
+            // emit upstream message if any
+            if let Some(msg) = upstream_buffer.next_pending_message().await? {
```

Comment:
Polling one message at a time may hurt barrier latency combining with the backoff above because barrier may not be polled in time.